client/jetpack-cloud/sections/partner-portal/hooks/use-issue-and-assign-licenses.tsx

```diff
@@ -0,0 +1,348 @@
+			// from what we recommend on the dashboard
+			if (
+				suggestedProducts.length > 0 &&
+				! containEquivalentItems( selectedProducts, suggestedProducts )
+			) {
+				dispatch(
+					recordTracksEvent(
+						'calypso_partner_portal_issue_multiple_licenses_changed_selection_after_dashboard_visit'
+					)
+				);
+			}
+
+			const selectedSiteId = selectedSite?.ID;
+			const fromDashboard = getQueryArg( window.location.href, 'source' ) === 'dashboard';
+
+			// If we need a payment method, redirect now to have the user enter one
+			if ( paymentMethodRequired ) {
+				const nextStep = addQueryArgs(
+					{
+						products: selectedProducts.join( ',' ),
+						...( selectedSiteId && { site_id: selectedSiteId } ),
+						...( fromDashboard && { source: 'dashboard' } ),
+					},
+					partnerPortalBasePath( '/payment-methods/add' )
+				);
+				return page( nextStep );
+			}
+
+			const issueLicenseResponses = await issueLicenses.issueLicenses( selectedProducts );
+
+			const issuedLicenses = issueLicenseResponses.filter(
+				( p ): p is PromiseFulfilledResult< APILicense > => p.status === 'fulfilled'
+			);
+			const issuedKeys = issuedLicenses.map( ( { value } ) => value.license_key );
+			const issuedProducts = issuedKeys
+				.map( ( key ) => {
+					const productSlug = key.split( '_' )[ 0 ];
+					return products?.data?.find( ( p ) => p.slug === productSlug );
+				} )
+				.filter( ( p ): p is APIProductFamilyProduct => p !== undefined );
+
+			// Exit early if we don't see any issued licenses matching a product we know
+			if ( issuedProducts.length === 0 ) {
+				return;
+			}
+
+			dispatch(
+				recordTracksEvent( 'calypso_partner_portal_multiple_licenses_issued', {
+					products: issuedProducts.join( ',' ),
+				} )
+			);
+
+			// If no site is selected, announce that licenses were issued;
+			// then, redirect to somewher more appropriate
+			if ( ! selectedSiteId ) {
+				const issuedProductNames: string[] = issuedProducts.map( ( { name } ) => name );
+				const issuedMessage = getLicenseIssuedMessage( issuedProductNames );
+				dispatch( successNotice( issuedMessage, { displayOnNextPage: true } ) );
+
+				// If this user has no sites, send them to the licenses listing page
+				if ( sitesCount === 0 ) {
+					return page.redirect( partnerPortalBasePath( '/licenses' ) );
+				}
+
+				// If they do have a site, send them to a page where they can assign
+				// the license(s) we just issued
+				const nextStep = addQueryArgs(
+					{
+						products: issuedKeys.join( ',' ),
+					},
+					partnerPortalBasePath( '/assign-license' )
+				);
+
+				return page.redirect( nextStep );
+			}
+
+			// If a specific site is already selected,
+			// let's assign the licenses we just issued to it
+			const assignLicenseResponses = await assignLicensesToSite.assignLicensesToSite( issuedKeys );
+
+			const assignedLicenses = assignLicenseResponses
+				.filter( ( p ): p is PromiseFulfilledResult< APILicense > => p.status === 'fulfilled' )
```

Comment:
I'm looking at the [corresponding code from trunk](https://github.com/Automattic/wp-calypso/blob/trunk/client/jetpack-cloud/sections/partner-portal/hooks.tsx#L423), and I don't see any error-handling code. Since I tried to preserve as much of the existing behavior as possible (to make the PR easier to review and test), I didn't address cases like this. It's something that could be added, but can you help me understand what that would look like?

Reply:
As you can see from the [corresponding code from trunk](https://github.com/Automattic/wp-calypso/blob/trunk/client/jetpack-cloud/sections/partner-portal/hooks.tsx#L423), we are not filtering the responses based on the status. We are using it [here](https://github.com/Automattic/wp-calypso/blob/trunk/client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-add-license-notification/index.tsx#L33)

Reply:
Agreed that we're referencing the `status` field -- I failed to fully explain. If the promise is rejected, TypeScript asserts that the `license` field is left undefined, which means rejected Promises would never make it into the array of assigned licenses.

As for the error message -- if I'm understanding you correctly, that is -- it's dispatched as part of the mutation itself on line 82.

Reply:
Ok, could we please get rid of this [piece of code](https://github.com/Automattic/wp-calypso/blob/trunk/client/jetpack-cloud/sections/agency-dashboard/sites-overview/site-add-license-notification/index.tsx#L33) in that case? I don't see any use of that still being there.